### PR TITLE
Add XML Parsing with `DocXmlElement`

### LINKED ISSUE #138: Support for table in TSDoc

### PR DESCRIPTION
**Pull request by suneettipirneni**
Closes #138 

This PR adds the ability for tsdoc to parse XML elements. For example the following input:

```
/*
 * <warning>
 *  This is a <bold>destructive</bold> operation.
 * </warning>
 */
```

Would yield this AST:
```
- Comment
  - Section
    - Paragraph
      - XmlElement
        * XmlStartTag_OpeningDelimiter="<"
        * XmlElement_Name="warning"
        * XmlStartTag_ClosingDelimiter=">"
        - PlainText
          * PlainText="this is a "
        - XmlElement
          * XmlStartTag_OpeningDelimiter="<"
          * XmlElement_Name="bold"
          * XmlStartTag_ClosingDelimiter=">"
          - PlainText
            * PlainText="destructive"
          * XmlEndTag_OpeningDelimiter="</"
          * XmlEndTag_Name="bold"
          * XmlEndTag_ClosingDelimiter=">"
        - PlainText
          * PlainText="operation."
        * XmlEndTag_OpeningDelimiter="</"
        * XmlEndTag_Name="warning"
        * XmlEndTag_ClosingDelimiter=">"
```

### Behavior

Unlike the current version of the parser, start and end tags alone aren't considered valid components. Take the following input:

```xml
<a>b</c>
```
Currently this input will result in a `DocHtmlStartTag` for `<a>`, `DocPlainText` for `b`, and `DocHtmlEndTag` for `</c>`. Each of these nodes viewed in isolation  are valid syntax, however, when parsed together as a unit they represent invalid XML. Thus, with the new XML parser this example would be considered erroneous:

```
Expecting closing tag name to match opening tag name, got "c" but expected "a"
```

As for its representation within the AST, the given example would not have anything resembling a `DocXmlElement` instead it would just be `DocErrorText`. Once again, this is because as a whole the input cannot be considered valid XML, therefore it's not constructed in any way as an XML node. 

That being said, I think my current implementation can be improved to provide more text even in erroneous code.

### Some Possible Changes

To align with the XML spec, XML nodes can have one of two nodes
- PlainText
- XmlElement

However, in writing the parser I was unsure about a possible third case: `SoftBreak`.

Softbreaks are valid children within XML elements. Currently softbreaks aren't parsed as individual nodes, but as a part of a `PlainText` node.

For example, the input:
```xml
<a>b
</a>
```

Would generate the following AST:
```
- XmlElement
        - PlainText
          * PlainText="b\n"
```

Alternatively the AST could be:
```
- XmlElement
        - PlainText
          * PlainText="b"
        - SoftBreak
          * SoftBreak="\n"
```

I'd love to hear feedback on which approach is preferred here.

Additionally, XML has [special escape characters](https://www.w3.org/TR/REC-xml/#dt-escape). Should we parse something like the following:
```
<a>1 is &lt; 100</a>
```

As

```
- XmlElement
        - PlainText
          * PlainText="1 is < 100"
```

I would also appreciate feedback on this as well.

### CODE REVIEW

**Comment by octogonz** [comment]
@suneettipirneni I finally got some time to look at this PR tonight -- it is really good! 👍👍 Very cool that you added lots of test coverage.  I need a little more time to study the logic in NdoeParser.ts, but overall this PR is on the right track.

**Comment by octogonz** [comment]
> However, in writing the parser I was unsure about a possible third case: `SoftBreak`.

According to the current design of the AST, I believe that `\n` is mostly forbidden in `DocPlainText`:
https://github.com/microsoft/tsdoc/blob/d8ce4aedcf64b5c948e7152ba02ac6c7bdedc9be/tsdoc/src/nodes/DocPlainText.ts#L49-L52

(I forget why this constraint was introduced. Maybe so that visitors that process `DocPlainText` do not have to worry about CRLF vs LF line ending differences?)

**Comment by octogonz** [comment]
> Additionally, XML has [special escape characters](https://www.w3.org/TR/REC-xml/#dt-escape). Should we parse something like the following:

Yes, in fact HTML entities are an important feature for representing arbitrary Unicode characters in an ASCII source file.  CommonMark [allows them basically anywhere](https://spec.commonmark.org/0.30/#entity-and-numeric-character-references) except for code spans, not just inside an XML element.  

But the problem is bigger than this PR.  For example, how best to represent an HTML entity inside an attribute value?  And should TSDoc allow entities inside other context such as `{@link ____}`?  

The problem is not super hard, but a rigorous specification of escaping syntaxes was one of the details that I thought we should figure out before writing the TSDoc spec.

File: tsdoc/src/parser/NodeParser.ts [hunk 1]
```diff
@@ -1826,105 +1821,268 @@ export class NodeParser {
...
-          marker,
-          'The HTML element has an invalid attribute: ',
+          attributeMarker,
+          'Invalid XML attribute: ',
           attributeNode
         );
       }
 
-      htmlAttributes.push(attributeNode);
+      xmlAttributes.push(attributeNode);
```
Comment (octogonz) [inline tsdoc/src/parser/NodeParser.ts:1870]: Not sure about HTML, but I'm pretty sure the XML spec forbids duplicated attributes (`<element a="1" a="2" />`). Should we validate that?

File: common/changes/@microsoft/tsdoc-config/feat-docxmlnodes_2022-08-03-21-34.json [hunk 2]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/tsdoc-config",
+      "comment": "Rename various properties to from HTML to XML.",
+      "type": "major"
```
Comment (octogonz) [inline common/changes/@microsoft/tsdoc-config/feat-docxmlnodes_2022-08-03-21-34.json:6]:
```suggestion
      "type": "minor"
```

For versions `<1.0.0` a MINOR bump is considered a breaking change. (and a SemVer range such as `^0.1.3` will not match `0.2.0`)

File: common/changes/@microsoft/tsdoc/feat-docxmlnodes_2022-08-03-21-34.json [hunk 3]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/tsdoc",
+      "comment": "Add support for XML parsing via `DocXmlElement`. Removes  `DocHtmlStartTag` and `DocHtmlEndTag` nodes.",
+      "type": "major"
```
Comment (octogonz) [inline common/changes/@microsoft/tsdoc/feat-docxmlnodes_2022-08-03-21-34.json:6]:
```suggestion
      "type": "minor"
```

File: tsdoc/src/parser/TSDocMessageId.ts [hunk 4]
```diff
@@ -339,44 +339,49 @@ export enum TSDocMessageId {
...
 
   /**
-   * Expecting an HTML tag starting with `</`.
+   * Expecting an XML tag starting with `</`.
    */
-  MissingHtmlEndTag = 'tsdoc-missing-html-end-tag',
+  MissingXmlEndTag = 'tsdoc-missing-xml-end-tag',
+
+  /**
+   * XML element tag is unterminated.
```
Comment (octogonz) [inline tsdoc/src/parser/TSDocMessageId.ts:370]: (Just a suggestion, open for discussion.) The word "unterminated" might be unclear to some readers, particularly if they weren't intending to write XML.

```suggestion
   * The corresponding XML end tag is missing.
```

This error gets a prefix, so the full message would be `Invalid XML element: The corresponding XML end tag is missing.`

If it's not too difficult, we could even include the name of the tag:  `The corresponding end tag "</abcd>" is missing.`  (substitute `abcd` with the actual name).

File: tsdoc/src/parser/TSDocMessageId.ts [hunk 5]
```diff
@@ -411,7 +416,12 @@ export enum TSDocMessageId {
   /**
    * The code span is missing its closing backtick.
    */
-  CodeSpanMissingDelimiter = 'tsdoc-code-span-missing-delimiter'
+  CodeSpanMissingDelimiter = 'tsdoc-code-span-missing-delimiter',
+
+  /**
+   * The XML start and end tag names are not the same.
+   */
+  XmlTagNameMismatch = 'tsdoc-xml-tag-name-mismatch'
```
Comment (octogonz) [inline tsdoc/src/parser/TSDocMessageId.ts:424]: Is this used? The error message that I see is `Expecting closing tag name "x" to match opening tag name "y"`

Comment (octogonz) [reply tsdoc/src/parser/TSDocMessageId.ts:424]: Looks like the tag is used but the comment is outdated

File: tsdoc/src/parser/NodeParser.ts [hunk 6]
```diff
@@ -1826,105 +1821,268 @@ export class NodeParser {
...
-    // NOTE: We don't read excerptParameters.separator here, since if there is any it
-    // will be represented as DocPlainText.
+    // Verify that the tag names are matching, if they aren't create a failure.
+    if (endTagNameExcerpt.toString() !== startTagNameExcerpt.toString()) {
+      return this._backtrackAndCreateErrorRange(
+        tokenReader,
+        endTagNameStartMarker,
+        endTagnameEndMarker,
+        TSDocMessageId.XmlTagNameMismatch,
+        `Expecting closing tag name "${endTagNameExcerpt.toString()}" to match opening tag name "${startTagNameExcerpt.toString()}"`
```
Comment (octogonz) [inline tsdoc/src/parser/NodeParser.ts:2009]: In the TSDoc Playground, the representation of this error is not quite right:

image

For the purposes of syntax highlighting and roundtripping, the parser tries to ensure that every input character corresponds to some AST excerpt, without any characters getting discarded.  

Some of the unit tests validate this property:
https://github.com/microsoft/tsdoc/blob/d8ce4aedcf64b5c948e7152ba02ac6c7bdedc9be/tsdoc/src/parser/__tests__/TestHelpers.ts#L177-L180

It raises a design question of how best to represent `</c>` in the AST?  

1. If it's too small of a fragment to be meaningful, then we could represent it as `ErrorText="</c>"`. 
2. For a richer representation, we could represent it as a loose `XmlElement` who is missing its opening excerpts. And maybe `<b>` would become a loose `XmlElement` who is missing its closing excerpts. (`<a>` does NOT need to be a child of `<b>` in this representation.)
3. An "intelligent" possibility would be to interpret `</c>` as the closing tag for `<b>`.  So `<b>` is a normal nesting element, just accompanied by an error message.

I don't have a strong opinion, but I suspect that 3 is maybe overengineering, and the decision of 1-vs-2 depends on how good we want the syntax highlighting to be, and how much time we have to spend on this PR. 🙂

Comment (suneettipirneni) [reply tsdoc/src/parser/NodeParser.ts:2024]: I'll try out option 2 and see if it's feasible, it seems like the most ideal solution in my opinion.

File: tsdoc/src/parser/TSDocMessageId.ts (hunk 4 above)
Comment (suneettipirneni) [reply tsdoc/src/parser/TSDocMessageId.ts]: I went ahead and added this with the corresponding tag name in the error: `The corresponding end tag "</abcd>" is missing.`

**Comment by suneettipirneni** [comment]
When I was fixing the duplicate attribute problem in the parser an idea popped into my head. I think we could improve the way we store attributes by using maps instead of arrays. From a developer perspective, if I want to access an attribute I'd usually already know the key for the value I want. Storing attributes in a map would allow developers to skip the boilerplate of using Array.find()

**Comment by octogonz** [comment]
> When I was fixing the duplicate attribute problem in the parser an idea popped into my head. I think we could improve the way we store attributes by using maps instead of arrays. From a developer perspective, if I want to access an attribute I'd usually already know the key for the value I want. Storing attributes in a map would allow developers to skip the boilerplate of using Array.find()

It's a helpful API, but I would suggest to expose it as a secondary API, something like `DocXmlElement.tryGetXmlAttribute(name): XmlAttribute | undefined`.  

We could convert the main model to be `xmlAttributes: Map<string, DocXmlAttribute>`, but this has some downsides:
- If we expose the `Map` directly, there's no way to validate items before they are added.
- Even if it is a `ReadonlyMap`, often you also want to validate keys before they are looked up.  ([one example](https://github.com/microsoft/tsdoc/blob/d8ce4aedcf64b5c948e7152ba02ac6c7bdedc9be/tsdoc/src/details/ModifierTagSet.ts#L53) from this repo)
- The order in which attributes appear can be syntactically useful to know, even if it has no semantic meaning. `Map` does not expose that very well.

In other words `Map` is the best way to implement lookups, but it is often not the best class to expose as public API contract.

**Comment by octogonz** [comment]
(let us know when this PR is ready for another look)

File: tsdoc/src/parser/NodeParser.ts (hunk 6 above)
Comment (suneettipirneni) [reply tsdoc/src/parser/NodeParser.ts:2024]: @octogonz I'm re-reading option 3 and just to be clear what you're proposing for that option is `ErrorText=<b>`, instead of setting the error text to `</c>`? If so, I think that may be easier than option 2.

Comment (octogonz) [reply tsdoc/src/parser/NodeParser.ts:2024]: To be clear, the parser has two ways of modeling syntax errors:
1. `DocErrorText` which will get syntax highlighted as red characters, plus an accompanying error message in the logs
2. a normal AST node such as `DocXmlElement`, whose properties are perhaps only partially initialized, plus an accompanying error message in the logs

The advantage of 2 is that syntax highlighting can be applied to the recognized characters, rather than having the entire span be red text.  The disadvantage of 2 is that if the AST node is too degenerate, then it is awkward for consumers of the API to handle, or perhaps completely incorrect (for example if we interpreted some stray characters `<y` as a `DocXmlElement`).

> just to be clear what you're proposing for that option is `ErrorText=<b>` instead of setting the error text to `</c>`? If so, I think that may be easier than option 2.

No, what I meant for option 3 was that excerpts `<b>` and `</c>` would belong to the same `DocXmlElement`.  I think it's not easier, because the some judgement calls are needed for less obvious such as `<a><b><c></c></a></b>` or `<a><b><c></c></a>`.

(If you want to save time, for an input like `<b><a></a></c>`, the least effort might be to generate `<b>` and `</c>` as `DocErrorText` -- it's not elegant, but we could always improve it later.)

I don't have a strong preference -- it's up to you.

Comment (suneettipirneni) [reply tsdoc/src/parser/NodeParser.ts]: Hi again @octogonz,

I actually went with option 3. The main reason I did this was because of the behaviors I observed when trying different XML parsers. Most of them properly figure out what's *supposed* to be the end tag for a given start tag. As for cases where the error non-obvious, I followed the errors I saw in the other parsers.

For example, 
```xml
<a><b><c></c></a></b>
```

Would yield the following AST:
```
	- XmlElement
        * XmlStartTag_OpeningDelimiter="<"
        * XmlElement_Name="a"
        * XmlStartTag_ClosingDelimiter=">"
        - XmlElement
          * XmlStartTag_OpeningDelimiter="<"
          * XmlElement_Name="b"
          * XmlStartTag_ClosingDelimiter=">"
          - XmlElement
            * XmlStartTag_OpeningDelimiter="<"
            * XmlElement_Name="c"
            * XmlStartTag_ClosingDelimiter=">"
            * XmlEndTag_OpeningDelimiter="</"
            * XmlEndTag_Name="c"
            * XmlEndTag_ClosingDelimiter=">"
          * XmlEndTag_OpeningDelimiter="</"
          * XmlEndTag_Name="a"
          * XmlEndTag_ClosingDelimiter=">"
        * XmlEndTag_OpeningDelimiter="</"
        * XmlEndTag_Name="b"
        * XmlEndTag_ClosingDelimiter=">"
```

Along with the following error in the log:
`Expected closing tag "<b/>" but received "<a/>" instead.`

A loose start tag will always try to find the next available *unmatched* end tag. If the end tag doesn't match a log error will be emitted and the AST for that particular element would be the start tag and then mismatched end tag as its end tag element.

`<a><b><c></c></a>`

Here `c` is a valid XML element, however, `b` is not because the next closest end tag is `</a>`. `a` is also not a valid XML element because the next closing tag closing tag `</c>` is matched with `<c>` already, the next closing tag after `</c>` is matched with `b`. This results in `a` not having a closing tag.

It can represented visually like so:
Untitled


Cases where the element is not large enough to create meaningful `DomXmlElement`, the text is simply considered `DocErrorText`.

With this I've been able to get mostly all of the valid syntax highlighting for loose nodes to work in various situations. I look forward to hearing your thoughts and feedback on this.

File: tsdoc/src/nodes/DocXmlElement.ts [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+  public tryGetXmlAttribute(name: string): DocXmlAttribute | undefined {
+    if (!DocXmlElement._isValidXmlName(name)) {
+      throw new Error(`'${name}' is not a valid XML attribute name.`);
+    }
+
+    return this._xmlAttributesByName.get(name);
+  }
+
+  private static _isValidXmlName(name: string): boolean {
+    return /^[a-zA-Z_][a-zA-Z0-9_.-]*$/.test(name);
```
Comment (octogonz) [inline tsdoc/src/nodes/DocXmlElement.ts:212]: `.` should not be allowed in attribute names.  At least this wasn't permitted in the previous doc, and the parser doesn't accept it because it relies on `StringChecks.explainIfInvalidXmlName()`.  

(The specs for XML/HTML/CommonMark do allow many other unusual characters, but for TSDoc we've taken a minimalist approach of only allowing syntaxes with a compelling justification. This is partly to make it more intuitive for humans to predict how TSDoc will get parsed, and partly to avoid difficult ambiguity/escaping edge cases caused by combining grammars of JSDoc, XML, and markdown.)

Comment (octogonz) [reply tsdoc/src/nodes/DocXmlElement.ts:212]: I will replace this RegExp with `StringChecks.explainIfInvalidXmlName()`

**Review by octogonz — APPROVED** [review]

**Comment by octogonz** [comment]
@suneettipirneni I greatly appreciate your patience with this long review. 😇🙏

I think this feature is ready to merge, except for a couple final questions:

## 1. Backwards compatibility

1. In testing, I realized that the upgraded `tsdoc-config` library cannot read old `tsdoc.json` files. This means that if several different tools are in use, they all need to be upgraded together.  I think we can make the upgrade a little easier by relaxing the schema/loader to accept the old setting names.  Then we can remove this backwards compatibility later, for example when we publish the 1.0 version of the library. 

I will do this work and push it to your branch.

## 2. Nested content

The handling of nested HTML content has changed with this PR.

**Old behavior:**
image

**New behavior:**
image

The behavior makes sense, because the meaning of a custom tag is user-defined. The content may be suitable to render verbatim (`<b>`, `<blockquote>`) but maybe not (`<style>`, `<meta>`). A reasonable conservative policy is to not render the content unless the tag is recognized.

Nonetheless, the example in the above screenshots is probably counterintuitive for end users.  I want to think a bit about how to handle this without adding nontrivial additional work to your PR.  If you have any ideas, let me know.

**Comment by octogonz** [comment]
> The handling of nested HTML content has changed with this PR.

🤔 Maybe what we need is a `tsdoc.json` setting that can specify: 

*"If your tool doesn't implement this XML element name, it's safe to process the element's content as if the element's tags were removed."*

Here's a more complicated example:

```ts
/**
 * <notebox>
 *   This is some <b>sample text</b>.
 *   <table-of-contents>
 *     <nav-node-name>My API Item</nav-node-name>
 *     <category>ui-controls</category>
 *   </table-of-contents>
 * </notebox>
 */
```

In this example, `<notebox>` and `<b>` can be safely ignored, whereas `<table-of-contents>`, `<nav-node-name>`, and `<category>` are data properties whose contents should NOT be rendered.

In `tsdoc.json` we would somehow define these element names and classify them as e.g. `"contentContainer"="true"`.  And then our default settings could apply this classification to common HTML elements such as `<a>`, `<b>`, `<i>`, etc.

**Comment by zelliott** [comment]
> The behavior makes sense, because the meaning of a custom tag is user-defined. The content may be suitable to render verbatim (`<b>`, `<blockquote>`) but maybe not (`<style>`, `<meta>`). A reasonable conservative policy is to not render the content unless the tag is recognized.

Am I correct in understanding that this is the behavior of the *TSDoc website* (i.e. whatever's actually rendering TSDoc), not TSDoc? I'm under the impression that TSDoc doesn't contain any rendering logic itself. So if the above behavior is unintuitive, wouldn't that mean fixing how the TSDoc website handles certain XML elements?

With the `tsdoc.json` setting proposal, I think you're suggesting that TSDoc can help inform a consumer which XML tags can have their contents rendered. There can be reasonable default settings (i.e. common HTML elements), but it can also be configurable. I suppose this is useful if the author of the TSDoc comments and the consumer rendering them are different people... is this a common use case?

**Comment by suneettipirneni** [comment]
> In tsdoc.json we would somehow define these element names and classify them as safe (e.g. contentContainer=true). And then our default settings could apply this classification to common HTML elements such as `<a>`, `<b>`, `<i>`, etc.

To some degree this option feels a bit out of place to me. My main dilemma is whether tsdoc should be making these distinctions or the renderers themselves. Renderers already make many implementation-specific decisions on the representation of a given doc node. Shouldn't they also be responsible for making the distinction of what's a `contentContainer` node and what isn't? 

I also fear this field may be confusing for consumers of TSDoc especially in cases where attributes play a key role. If an `<a>` element is marked as a `contentContainer` and my custom markdown emitter automatically outputs the plaintext of any content-container element I could inadvertently skip over the `href` attribute. 

```ts
for (const node of nodes) {
	switch (node.kind) {
		case DocNodeKind.XmlElement:
			const xmlNode = node as DocXmlElement;
			if (xmlNode.contentContainer) {
				const text = // Logic to resolve the inner plain text(s) and convert them to strings ... ;
				output.append(text);
				break;
			}

			// Otherwise render recognized XML elements
	}
}
```

While this code is ultimately a logic error by the programmer (you could argue that custom rendered elements should be checked first) it may be caused by the somewhat valid assumption that `a` *is* a `contentContainer` since it only contains text as child elements. This could also become an issue if there's lack of shared knowledge between the programmers writing the docs and the programmers creating emitters.

**Comment by octogonz** [comment]
> Am I correct in understanding that this is the behavior of the _TSDoc website_ (i.e. whatever's actually rendering TSDoc), not TSDoc? 

Yes. Here's some examples of various possible "renderers" of TSDoc:

- The [TSDoc playground](https://tsdoc.org/play) shown in those screenshots
- An API reference website generated using [API Documenter](https://api-extractor.com/pages/setup/generating_docs/)
- A custom web app that displays API documentation by parsing TSDoc from `.api.json` files produced by API Extractor
- It doesn't need to be a website: for example, a VS Code extension that shows rich tooltips by parsing TSDoc
- It doesn't even need to render anything: for example, a lint rule that counts how many English words are in your TSDoc summary

Importantly, multiple different tools may need to process the same code comment, which is why ***interoperability*** between tools is a key requirement for TSDoc.

> I'm under the impression that TSDoc doesn't contain any rendering logic itself. 

Yes.  But TSDoc does provide [a config file](https://tsdoc.org/pages/packages/tsdoc-config/) that allows users to define custom tags (both JSDoc tags and XML tags).  This configuration enables the parser to report syntax errors about undefined tags, and (for JSDoc tags at least) it distinguishes **"undefined"** tags versus **"unsupported"** tags:

- undefined =  not a recognized tag in the TSDoc dialect that was configured by the person who wrote the comment
- unsupported = defined but not implemented by a particular tool that may be inspecting the file.  

TSDoc's own [standard tags](https://github.com/microsoft/tsdoc/blob/main/tsdoc/src/details/StandardTags.ts) are all "defined" by default, whereas most tools will only "support" a subset of them.

> My main dilemma is whether tsdoc should be making these distinctions or the renderers themselves. Renderers already make many implementation-specific decisions on the representation of a given doc node. Shouldn't they also be responsible for making the distinction of what's a `contentContainer` node and what isn't?

Well, the original question was:

*"If a tool doesn't implement an XML element, what should it do?"*

The simple, safe answer is to ignore the element and all of its content.  

But suppose I invent a custom tag `<glossary-word>` and use it to mark up sentences like `This is a <glossary-word>book<glossary-word>.`  If a tool doesn't support this tag, deleting the content would be a bad experience.  Do we really need to release new versions of every tool to ensure that `<glossary-word>`'s content can be rendered safely?
